File: cython_pkg/short_link.py
from abc import ABC

import cython

# 64 chars to encode 6 bits
_ARRAY = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_~'
_ARRAY_MAP = {c: i for i, c in enumerate(_ARRAY)}
_ARRAY_MAP['@'] = _ARRAY_MAP['~']  # backwards compatibility


class ShortLink(ABC):
    @staticmethod
    def encode(lon: cython.double, lat: cython.double, z: cython.int) -> str:
        '''
        Encode a coordinate pair and zoom level into a short link string.
        '''

        x: cython.int = int((lon + 180) * 11930464.711111112)  # (2 ** 32) / 360
        y: cython.int = int((lat + 90) * 23860929.422222223)  # (2 ** 32) / 180
        c: cython.int = 0
        i: cython.int

        for i in range(31, -1, -1):
            c = (c << 2) | (((x >> i) & 1) << 1) | ((y >> i) & 1)

        d: cython.int = z + 8 // 3
        r: cython.int = z + 8 % 3

        if r > 0:  # ceil instead of floor
            d += 1

        str_list = ['-'] * (d + r)

        for i in range(d):
            digit: cython.int = (c >> (58 - 6 * i)) & 0x3F
            str_list[i] = _ARRAY[digit]

        return ''.join(str_list)

    @staticmethod
    def decode(s: str) -> tuple[cython.double, cython.double, cython.int]:
        '''
        Decode a short link string into a coordinate pair and zoom level.
        '''

        x: cython.int = 0
        y: cython.int = 0
        z: cython.int = len(s) * 3
        z_offset: cython.int = 0

        for c in s:
            t: cython.int = _ARRAY_MAP.get(c, -1)

            if t == -1:
                z_offset -= 1
                continue

            for _ in range(3):
                x = (x << 1) | ((t >> 5) & 1)
                y = (y << 1) | ((t >> 4) & 1)
                t <<= 2

        x <<= (32 - z)
        y <<= (32 - z)

        return (
            (
                x * 8.381903171539307e-08  #360 / (2 ** 32)
            ) - 180,
            (
                y * 4.190951585769653e-08  # 180 / (2 ** 32)
            ) - 90,
            z - 8 - (z_offset % 3)
        )

File: app/static/ts/_shortlink.ts
import { mod } from "./_utils"
import type { LonLatZoom } from "./leaflet/_map-utils"

/**
 * Encode a coordinate and zoom level to a short link code
 * @example
 * shortLinkEncode(5.123, 10.456, 17)
 * // => "wF7ZdNbjU-"
 */
export const shortLinkEncode = ({ lon, lat, zoom }: LonLatZoom): string => {
    const z = (zoom | 0) + 8
    const d = Math.ceil((z + 8) / 3)
    const r = (z + 8) % 3

    const x = BigInt((mod(lon + 180, 360) * 11930464.711111112) | 0) // (2 ** 32) / 360
    const y = BigInt(((lat + 90) * 23860929.422222223) | 0) // (2 ** 32) / 180

    let c = 0n
    for (const mask of bitMasks) {
        // noinspection JSBitwiseOperatorUsage
        c = (c << 2n) | (x & mask ? 2n : 0n) | (y & mask ? 1n : 0n)
    }

    const buffer = new Array(d + r)
    for (let i = 0; i < d; i++) {
        buffer[i] = code[Number((c >> BigInt(58 - 6 * i)) & 0x3fn)]
    }
    for (let i = d; i < buffer.length; i++) {
        buffer[i] = "-"
    }

    return buffer.join("")
}

/** 64 chars to encode 6 bits */
const code = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_~"
const bitMasks = Object.freeze([
    2147483648n,
    1073741824n,
    536870912n,
    268435456n,
    134217728n,
    67108864n,
    33554432n,
    16777216n,
    8388608n,
    4194304n,
    2097152n,
    1048576n,
    524288n,
    262144n,
    131072n,
    65536n,
    32768n,
    16384n,
    8192n,
    4096n,
    2048n,
    1024n,
    512n,
    256n,
    128n,
    64n,
    32n,
    16n,
    8n,
    4n,
    2n,
    1n,
]) // Array.from({ length: 32 }, (_, i) => 1n << BigInt(31 - i))
